go-controller/pkg/node/default_node_network_controller.go

```diff
@@ -760,6 +794,112 @@ func (nc *DefaultNodeNetworkController) Start(ctx context.Context) error {
+	// OVN/OVS flows are intact.
+	// STEP1: ovnkube-node start's up in remote zone and sets the "k8s.ovn.org/zone-name" above.
+	// STEP2: We delay the flip of connection for ovnkube-node(ovn-controller) to the new remote SBDB
+	//        until the new remote ovnkube-controller has finished programming all the K8s core objects
+	//        like routes, services and pods. Until then the ovnkube-node will talk to legacy SBDB.
+	// STEP3: Once we get the signal that the new SBDB is ready, we set the "k8s.ovn.org/remote-zone-migrated" annotation
+	// STEP4: We call setDBAuth to now point to new SBDB
+	// STEP5: Legacy ovnkube-master sees "k8s.ovn.org/remote-zone-migrated" annotation on this node and now knows that
+	//        this node has remote-zone-migrated successfully and tears down old setup and creates new IC resource
+	//        plumbing (takes 80ms based on what we saw in CI runs so we might still have that small window of disruption).
+	if config.OVNKubernetesFeature.EnableInterconnect && sbZone != types.OvnDefaultZone && !util.HasNodeMigratedZone(node) { // so this should be done only once in phase2 (not in phase1)
+		klog.Info("Upgrade Hack: Interconnect is enabled")
+		nc.wg.Add(1)
+		go func() {
+			defer nc.wg.Done()
+			var err1 error
+			start := time.Now()
+			err = wait.PollUntilContextTimeout(context.Background(), 500*time.Millisecond, 300*time.Second, true, func(ctx context.Context) (bool, error) {
+				nodes, err := nc.Kube.GetNodes()
+				if err != nil {
+					err1 = fmt.Errorf("upgrade hack: error retrieving node %s: %v", nc.name, err)
+					return false, nil
+				}
+
+				// we loop through all the nodes in the cluster and ensure ovnkube-controller has finished creating the LRSR required for pod2pod overlay communication
+				for _, node := range nodes.Items {
+					if nc.name != node.Name {
+						nodeSubnets, err := util.ParseNodeHostSubnetAnnotation(&node, types.DefaultNetworkName)
+						if err != nil {
+							err1 = fmt.Errorf("unable to fetch node-subnet annotation for node %s: err, %v", node.Name, err)
+							return false, nil
+						}
+						for _, nodeSubnet := range nodeSubnets {
+							klog.Infof("Upgrade Hack: node %s, subnet %s", node.Name, nodeSubnet)
+							if !checkOVNSBNodeLRSR(nodeSubnet.String()) {
+								err1 = fmt.Errorf("upgrade hack: unable to find LRSR for node %s", node.Name)
+								return false, nil
+							}
+						}
+					}
+				}
+				// we loop through all existing services in the cluster and ensure ovnkube-controller has finished creating LoadBalancers required for services to work
+				services, err := nc.watchFactory.GetServices()
+				if err != nil {
+					err1 = fmt.Errorf("upgrade hack: error retrieving the services %v", err)
+					return false, nil
+				}
+				for _, s := range services {
+					// don't process headless service
+					if !util.ServiceTypeHasClusterIP(s) || !util.IsClusterIPSet(s) {
+						continue
+					}
+					if !lbExists(s.Namespace, s.Name) {
```

Comment:
lbExists just lists all load balancers in SBDB, so you continuously call ovn-sbctl for every service here. I think it would be easier to just do 1 sbctl call, then check for all the services in the output.

Reply:
yeah very true, much cleaner and less noise and possibly less calls - besides services doesn't really change much in the cluster during network rollout so should be a finite set of those services.